cohortextractor/cohortextractor.py

```diff
@@ -220,6 +223,78 @@ def _increment_date(date, period):
 
 
+def generate_measures(output_dir, selected_study_name=None, skip_existing=False):
+    preflight_generation_check()
+    study_definitions = list_study_definitions()
+    if selected_study_name and selected_study_name != "all":
+        for study_name, suffix in study_definitions:
+            if study_name == selected_study_name:
+                study_definitions = [(study_name, suffix)]
+                break
+    for study_name, suffix in study_definitions:
+        print(f"Generating measure for {study_name}...")
+        _generate_measures(output_dir, study_name, suffix, skip_existing=skip_existing)
+
+
+def _generate_measures(output_dir, study_name, suffix, skip_existing=False):
+    print("Running. Please wait...")
+    measures = load_study_definition(study_name, value="measures")
+    files = glob.glob(f"{output_dir}/input{suffix}*.csv")
+    if not files:
+        print(
+            "No matching output files found. You may need to first run:\n"
+            "  cohortextractor generate_cohort --index-date-range ..."
+        )
+        return
+    for file in files:
+        date = _get_date_from_filename(file)
+        patient_df = None
+        for measure in measures:
+            output_file = f"{output_dir}/measure_{measure.id}_{date}.csv"
+            if skip_existing and os.path.exists(output_file):
+                print(f"Not generating pre-existing file {output_file}")
+                continue
+            # We do this lazily so that if all corresponding output files
+            # already exist we can avoid loading the patient data entirely
+            if patient_df is None:
+                patient_df = _load_csv_for_measures(file, measures)
+            measure_df = patient_df[
+                [measure.numerator, measure.denominator, measure.group_by]
+            ]
+            measure_df = measure_df.groupby(measure.group_by).sum()
+            measure_df["value"] = (
+                measure_df[measure.numerator] / measure_df[measure.denominator]
+            )
+            measure_df.to_csv(output_file)
+            print(f"Created measure output at {output_file}")
+
+
+def _get_date_from_filename(filename):
+    match = re.search(r"_(\d\d\d\d\-\d\d\-\d\d)\.csv$", filename)
+    return datetime.date.fromisoformat(match.group(1))
+
+
+def _load_csv_for_measures(file, measures):
+    """
+    Given a CSV file name and a list of measures, load the file into a Pandas
+    dataframe with types as appropriate for the supplied measures
+    """
+    numeric_columns = set()
+    other_columns = set()
+    for measure in measures:
+        numeric_columns.update([measure.numerator, measure.denominator])
+        other_columns.add(measure.group_by)
+    # This is a special column which we don't load form the CSV but whose value
+    # is always set to 1 for every row
+    numeric_columns.discard("population")
+    dtype = {col: "str" for col in other_columns}
```

Comment:
Given `other_columns` is always the `group_by` (right? in which case, in passing, should the variable name be `group_by_columns`?) and therefore *should* be of low cardinality, I think the appropriate dtype is probably `categorical` - would save lots of memory.